Gawron/ZPI-39-ZPI-15-BE-15-kalendarz-z-dostepnościa-opiekuna

Gawron97 opened the pull request:
Dodanie filtrowania po dostępności opiekuna.
Dodanie sortowania po liczbie pokrywających się dni dostępności. Jeśli sortujemy po tym atrybucie, to ma ono największą wagę, dlatego, że sortowanie po liczbie pokrywających się dni dostępności wykonywane jest na końcu działania.

Gawron97 committed:
add filtering by availability
returns caretakers which has at least one offer where availability match for at least one day

Gawron97 committed:
add tests for filtering by offer availability

Gawron97 committed:
merge main

Gawron97 committed:
calculating availability days match and sort by this

Gawron97 committed:
add test for calculating availabilityDaysMatch

Gawron97 committed:
fix test

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/CaretakerService.java:
```diff
@@ -40,12 +50,105 @@ public class CaretakerService {
+    public Page<CaretakerDTO> getCaretakers(Pageable pageable, boolean hasSortByAvailabilityDaysMatch,
+                                            Sort.Direction direction, CaretakerSearchCriteria filters,
+                                            Set<OfferFilterDTO> offerFilters) {
         Specification<Caretaker> spec = CaretakerSpecificationUtils.toSpecification(filters, offerFilters);
```
Dałbym każdy parametr w nowej linii

lukcyn commented on src/main/java/com/example/petbuddybackend/controller/CaretakerController.java:
```diff
@@ -47,8 +49,16 @@ public Page<CaretakerDTO> getCaretakers(
+        if(pagingParams.getSortBy().contains("availabilityDaysMatch")) {
+            hasSortByAvailabilityDaysMatch = true;
+            pagingParams.getSortBy().remove("availabilityDaysMatch");
+        }
```
Imo to mogłoby zostać w serwisie

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/CaretakerService.java:
```diff
@@ -40,12 +50,105 @@ public class CaretakerService {
+                    filteredCaretakers,
+                    direction
+            );
+        }
```
My chcemy takie sortowanie? Bo wynikiem tego będzie, że posortujemy opiekunów w ramach danego Page'a, a nie w ramach wszystkich opiekunów. Wyjdzie z tego taka sytuacja, ze na pierwszej stronie będzie opiekun co ma tylko 1 dzień dostępny a na następnej taki, co ma wszyskie dni dostępne

Gawron97 replied:
aa rzeczywiscie, wszystkich trzeba posortowac

lukcyn commented:
Zastanawia mnie czy by tu jakieś query nie wchodziło w gre ze względu na stopień skomplikowania tego GETa

Gawron97 committed:
change searching caretakers by availability
get caretaker if for all offers at least one availability range is overlapped
all days of availability range from filter must be overlapped by one of caretaker offer availability range

Gawron97 committed:
remove availabilityDaysMatch

Gawron97 committed:
refactor CaretakerSpecificationUtils

lukcyn commented on src/test/java/com/example/petbuddybackend/testutils/PersistenceUtils.java:
```diff
@@ -105,8 +107,10 @@ public static AppUser addAppUser(AppUserRepository appUserRepository, AppUser ap
     }
 
     public static Offer addComplexOffer(Caretaker caretaker, Animal animal, List<AnimalAttribute> animalAttributes,
-                                        BigDecimal price, List<AnimalAmenity> animalAmenities, OfferRepository offerRepository) {
-        Offer offer = addComplexMockOfferForCaretaker(caretaker, animal, animalAttributes, price, animalAmenities);
+                                        BigDecimal price, List<AnimalAmenity> animalAmenities,
+                                        Set<Availability> availabilities, OfferRepository offerRepository) {
+        Offer offer = addComplexMockOfferForCaretaker(caretaker, animal, animalAttributes, price, animalAmenities,
```
Tu też bym porozbijał parametry do osobnych linii

lukcyn commented on src/main/java/com/example/petbuddybackend/service/user/CaretakerService.java:
```diff
@@ -40,11 +50,11 @@ public class CaretakerService {
 
     @Transactional(readOnly = true)
-    public Page<CaretakerDTO> getCaretakers(Pageable pageable, CaretakerSearchCriteria filters, Set<OfferFilterDTO> offerFilters) {
+    public Page<CaretakerDTO> getCaretakers(Pageable pageable, CaretakerSearchCriteria filters,
+                                            Set<OfferFilterDTO> offerFilters) {
```
Porozbijałbym parametry do osobnych linii

lukcyn commented on src/main/java/com/example/petbuddybackend/utils/specification/CaretakerSpecificationUtils.java:
```diff
@@ -276,4 +285,51 @@ private static Predicate amenityExists(Subquery<Long> offerSubquery, CriteriaBui
 
     }
+
+    private static Predicate availabilityMatchForOffer(Root<Caretaker> root, CriteriaQuery<?> query,
+                                                       CriteriaBuilder cb, OfferFilterDTO offerFilter) {
```
Porozbijałbym parametry do osobnych linii

lukcyn reviewed (approved):
Tylko drobne sugestie żeby łatwiej się czytało, a tak to wygląda git

Gawron97 committed:
refactor as review
add more tests for getCaretakers
